### PR TITLE
Nakamoto: Validate stacker signature of a Nakamoto Block

### PR DESCRIPTION
**Pull request by jferrant**
### Description

Add functionality to check if a Nakamoto block has been correctly signed with the Signers (Stackers) WSTS signature associated with the aggregate public key. 

### Applicable issues
- fixes https://github.com/stacks-network/stacks-blockchain/issues/3927

### CODE REVIEW

File: stackslib/src/chainstate/burn/db/sortdb.rs
Comment (obycode) [inline stackslib/src/chainstate/burn/db/sortdb.rs:1970]: Could this be an `assert_eq!` instead?

**Review by obycode — APPROVED** [review]
LGTM! I just had one minor suggestion. 
May the Schwartz be with you.

File: stackslib/src/chainstate/burn/db/sortdb.rs
Comment (kantai) [inline stackslib/src/chainstate/burn/db/sortdb.rs:1974]: Not a huge issue, but I think in a case with nested `map()` invocations, it's probably more clear to flatten the nesting use the try syntax `?` instead:

```suggestion
        let Some(pk_string) = self.get_indexed(sortition_id, &db_keys::pox_stackers_aggregate_public_key())?
        else {
              return Ok(None)
        };
        let pk_byte_vec = hex_bytes(&pk_string).expect("CORRUPTION: bad stackers aggregate public key. Invalid hex string.");
        let mut pk_bytes: [u8; 33] = [0; 33];
        if pk_bytes.len() != pk_byte_vec.len() {
             panic!("CORRUPTION: bad stackers aggregate public key. Expected {} bytes, got {} bytes", x_bytes.len(), x.len());
        }
        pk_bytes.copy_from_slice(&pk_byte_vec[0..33]);
        Ok(Point::try_from(&Compressed::from(pk_bytes)).expect("CORRUPTION: bad stackers aggregate public key"))
```

**Review by kantai — APPROVED** [review]
LGTM -- only had a minor comment

**Review by jcnelson — CHANGES_REQUESTED** [review]
I don't think we should be using `MessageSignature` to encapsulate a Schnorr signature. I was using that as a placeholder for a Schnorr-specific signature type.

If you can instead create a Schnorr signature container, and use that in place of `MessageSignature` in the block header, then I'll approve :+1:

File: stacks-common/src/util/secp256k1.rs
Comment (jcnelson) [inline stacks-common/src/util/secp256k1.rs:740]: lol

File: stackslib/src/chainstate/burn/db/sortdb.rs
Comment (jcnelson) [reply stackslib/src/chainstate/burn/db/sortdb.rs:1974]: Yeah, it's generally a good idea to avoid excessive indentations

**Comment by jferrant** [comment]
> I don't think we should be using `MessageSignature` to encapsulate a Schnorr signature. I was using that as a placeholder for a Schnorr-specific signature type.
> 
> If you can instead create a Schnorr signature container, and use that in place of `MessageSignature` in the block header, then I'll approve 👍

Done. Just working on adding "aggregate_public_key" to the RewardCycleInfo

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (jferrant) [inline stackslib/src/chainstate/stacks/boot/mod.rs:1157]: Fix typo in call

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/mod.rs:1157]: Is there a `get-aggregate-public-key` function in `pox-4`?  If not, can you add it?

Also, I don't think this is ever going to be optional in PoX-4.  Instead, this method should just return `Result<Point, Error>`.

Comment (jferrant) [reply stackslib/src/chainstate/stacks/boot/mod.rs:1157]: This could be optional for a given reward cycle as signers may not have set it yet no? In theory they should have set it BEFORE the start of their reward cycle, but they may fail to do so before the start.

**Comment by jferrant** [comment]
I am also inclined in this PR to update stacker_signature and all references to it to signers_signature as technically stackers may have offloaded their responsibilities to some other signer operator. Does anyone have objections to this? @jcnelson @kantai

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/mod.rs]: > In theory they should have set it BEFORE the start of their reward cycle, but they may fail to do so before the start and be still executing a DKG round AFTER the start of a reward cycle.

Ah, yes, you're right. This has to be optional. The signers can't compute the aggregate public key until after DKG, so there's going to be a delay between calculating the reward set and setting the aggregate public key in `pox-4`.

However, I think you'll need to update `get_nakamoto_reward_cycle_info()` to return `None` if the aggregate public key is not known.  We should keep trying to compute the reward set until we have an aggregate public key.

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/mod.rs]: > but they may fail to do so before the start and be still executing a DKG round AFTER the start of a reward cycle.

As of right now, if this happens, then the chain will freeze.  There's no way for the Stacks chain state to advance if the next reward cycle begins without the new aggregate public key. 

This could be addressed if there was a Stacks-on-Bitcoin way to set the aggregate public key, however.  I'll open a discussion for it in the SIP repo.

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/mod.rs]: https://github.com/stacksgov/sips/discussions/162

**Comment by jcnelson** [comment]
Sounds good to me :+1:

**Comment by jferrant** [comment]
Waiting on https://github.com/stacks-network/stacks-core/pull/4009 to be merged
Waiting on the merge of Developer into Next
Will rebase once that is done

Adding testing now.

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jferrant) [inline stackslib/src/chainstate/nakamoto/coordinator/tests.rs:388]: This test is currently failing. Seems to be failing on third iteration of the loop. 

```
WARN [1700578532.327215] [stackslib/src/chainstate/burn/db/sortdb.rs:2038] [chainstate::nakamoto::coordinator::tests::test_simple_nakamoto_coordinator_10_tenures_10_blocks] reward start height 41 for reward cycle 8 does not have a sorition from 54538e85cabb2462772bb668b6a8d1e967d837c1060788fd289a295c93424c49
```

Jude has discovered that it is off by one, looking at start height 41 when it should be at height 40. Why this off by one only happens in this test on the third iteration is unknown still.

Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/coordinator/tests.rs:388]: Found part of the problem: find_prepare_phase_sortitions was incorrectly adding the sortition ID of one BEFORE the prepare phase. Still failing at replay (I think because it tries to access the aggregate public key when it is not set)

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/coordinator/tests.rs:437]: Why do we need to set the aggregate public key multiple times (e.g. in each block)?

File: stackslib/src/chainstate/nakamoto/tests/node.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/tests/node.rs:73]: We use the word `comms` elsewhere to refer to "communication channels".  Can this member be renamed to something like `signer_commits` to remove ambiguity?

File: stackslib/src/chainstate/stacks/boot/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/mod.rs:1138]: Nit: `u128::from(reward_cycle)`

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:1332]: You don't need to remove this _per se_; you should instead just make it a private function, and have the node itself call it at the start of each PoX-4 reward cycle. You don't have to do this in this PR, but it does need to happen prior to Neon because this function should not be publicly reachable.

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:1332]: Please open a Neon-tagged issue for the above :pray:

**Review by jcnelson — APPROVED** [review]
LGTM; just minor nits to address before merging.  Thanks @jferrant for powering through this!

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/coordinator/tests.rs:437]: Oh I see, we still need to add a way for the node itself to do this as part of processing a reward cycle.

Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/coordinator/tests.rs:437]: It probably does not need to be EVERY block. I just decided lets do overkill to avoid missing processing it during replay. In theory this should not be needed as we simulate signer's submitting their votes across a reward cycle rather than a straight "set-aggregate-public-key" call. (this call should be removed once we have the ability to vote correctly)

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/mod.rs:1447]: Will do!

File: stackslib/src/chainstate/nakamoto/tests/node.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/tests/node.rs:73]: Will do!

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jferrant) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:1332]: I have created and also adding as comment in code: https://github.com/stacks-network/stacks-core/issues/4101

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/mod.rs:1427]: `canonical_block_header` is the block that the `aggregate_public_key` will be fetched at, which isn't necessary the "canonical chain tip" -- can you rename this variable to something like `fetch_at_block`?

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/mod.rs:1426]: `header` doesn't really tell me much other than the type of the variable, perhaps, `for_block` ?

File: stackslib/src/net/relay.rs
Comment (kantai) [inline stackslib/src/net/relay.rs:717]: I think this shouldn't be the `canonical_block_header`. It should be the last stacks block of the prepare phase or the parent of the stacks block being considered.

I don't think this needs to block this PR, though, so I opened an issue (#4109) -- can you just add a todo comment, with the issue number in it?

**Review by kantai — APPROVED** [review]
LGTM -- I think we should just flag a relevant issue for the block that the agg key is fetched from, I don't think it needs to block this PR.

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/mod.rs:1427]: I think that, specifically, this needs to be at least as high as the last block in the last reward cycle.

File: stackslib/src/net/relay.rs
Comment (jferrant) [reply stackslib/src/net/relay.rs:717]: Will do!

**Pull request merged**
